src/fiber_prompt/wasmfx/imports.wat.pp

```diff
@@ -0,0 +1,163 @@
+
+    (local.get $func_index)
+    (local.get $trampoline_index)
+    (cont.new $ct3 (ref.func $wasmfx_entry_trampoline))
+    (cont.bind $ct3 $ct1)
+    (local.set $cont)
+
+    (table.set $conts (local.get $cont_index) (local.get $cont))
+  )
+
+  (func $indexed_resume (export "wasmfx_indexed_resume_with")
+    (param $cont_index i32)
+    (param $arg i32)
+    (param $result_ptr i32)
+    (result i32)
+    (local $k (ref $ct1))
+    (local $old_shadow_sp i32)
+
+#ifdef FIBER_WASMFX_PRESERVE_SHADOW_STACK
+    (local.set $old_shadow_sp (global.get $sstack_ptr))
+    ;; The following 5 instructions are equivalent to the following C code
+    ;; (inlined here until we have wasm-opt):
+    ;; __shadow_stack_ptr = sstack_current_ptrs[cont_index]
+
+    (i32.load (global.get $sstack_current_ptrs_addr))
+    (i32.mul (i32.const 4) (local.get $cont_index))
+    (i32.add)
+    (i32.load)
+    (global.set $sstack_ptr)
+#endif
+
+    (block $handler (result i32 (ref $ct1) )
+      (block $on_error
+        ;; Retrieve the continuation and test for null
+        (local.set $k (br_on_null $on_error
+           (table.get $conts (local.get $cont_index))))
+        ;; set active_fiber to store table index for names
+        (global.set $active_fiber (local.get $cont_index))
+        ;; resume the continuation
+        (resume_with $ct1 (on $yield $handler) (local.get $arg) (local.get $k))
+        (i32.store (local.get $result_ptr) (i32.const 0)) ;; FIBER_OK
+        (table.set $conts (local.get $cont_index) (ref.null $ct1))
+
+#ifdef FIBER_WASMFX_PRESERVE_SHADOW_STACK
+        ;; No need to save $sstack_ptr as continuation finished
+        (global.set $sstack_ptr (local.get $old_shadow_sp))
+#endif
+
+        (return) ;; returns the value put on stack by resume
+      ) ;; continuation is null
+      (i32.store (local.get $result_ptr) (i32.const 2)) ;; FIBER_ERROR
+      (return (i32.const -1))
+    )
+    (local.set $k)
+
+    ;; stash continuation aside
+    (table.set $conts (local.get $cont_index) (local.get $k))
+
+    (i32.store (local.get $result_ptr) (i32.const 1)) ;; FIBER_YIELD
+
+#ifdef FIBER_WASMFX_PRESERVE_SHADOW_STACK
+    ;; The following 5 instructions are equivalent to the following C code
+    ;; (inlined here until we have wasm-opt):
+    ;; sstack_current_ptrs[cont_index] = __shadow_stack_ptr
+    (i32.load (global.get $sstack_current_ptrs_addr))
+    (i32.mul (i32.const 4) (local.get $cont_index))
+    (i32.add)
+    (global.get $sstack_ptr)
+    (i32.store)
+
+    (global.set $sstack_ptr (local.get $old_shadow_sp))
+#endif
+
+    ;; return suspend payload
+    (return)
+  )
+  
+  (func $suspend_to (export "wasmfx_suspend_to") (param $arg i32) (result i32)
+   (local $updated_name (ref null $ht))
+   (suspend_to $ht $yield (local.get $arg) (table.get $names (global.get $active_fiber)))
+   (local.set $updated_name)
+   (table.set $names (global.get $active_fiber) (local.get $updated_name))
+   (return)
+  )
```

Comment:
I think this can be made more succinct.
```suggestion
  (func $suspend_to (export "wasmfx_suspend_to") (param $arg i32) (result i32)
   (suspend_to $ht $yield (local.get $arg) (table.get $names (global.get $active_fiber)))
   (table.set $names (global.get $active_fiber))
  )
```